### PR TITLE
Enable labels for GKE clusters and add tests

### PR DESCRIPTION
**Pull request by ksonney**
This adds support for labeling to GKE clusters, as well as some misc readability updates to the tests. 

Closes: #8529 

DRAFT TBD: 
- [x] Needs to invalidate and replace the cache entry for the GKE cluster or additional tag operations on the cluster will fail due to a mismatched Label Fingerprint.
- [x] Awaiting addition to the Elastic CNCF Organization CLA
- [x] Working tests

### CODE REVIEW

File: tools/c7n_gcp/tests/test_gke.py
Comment (ajkerrigan) [inline tools/c7n_gcp/tests/test_gke.py:76]: Nice use of `skip` here to show progress/intent, very helpful as a reader :+1:

Comment (ksonney) [reply tools/c7n_gcp/tests/test_gke.py:76]: Thanks, until we can get a clean recording, that's about the only thing we can do - otherwise the test will fail.

**Comment by ksonney** [comment]
>     * [ ]  Needs to invalidate and replace the cache entry for the GKE cluster or additional tag operations on the cluster will fail due to a mismatched Label Fingerprint.

This also happens with instances (and I suspect other resources) when I was testing. I'm kind of surprised that it hasn't been raised as an issue in the past.

**Comment by kapilt** [comment]
> > ```
> > * [ ]  Needs to invalidate and replace the cache entry for the GKE cluster or additional tag operations on the cluster will fail due to a mismatched Label Fingerprint.
> > ```
> 
> 
> This also happens with instances (and I suspect other resources) when I was testing. I'm kind of surprised that it hasn't been raised as an issue in the past.

discussed in community meeting, I think we just have to try/except refetch the resource to get the current fingerprint and retry, to handle stale.

**Comment by ksonney** [comment]
> discussed in community meeting, I think we just have to try/except refetch the resource to get the current fingerprint and retry, to handle stale.

Both gke and compute now re-fetch the fingerprint and retry on a fingerprint error. There is probably a more elegant way to do this, but this much works.

**Comment by ksonney** [comment]
We have working tests, so taking this out of draft status and marking as ready for review!

File: tools/c7n_gcp/tests/test_gke.py
Comment (ajkerrigan) [inline tools/c7n_gcp/tests/test_gke.py:89]: This looks good, though I think if we want to work the fingerprint mismatch logic in this test we need to:

- Enable caching (tests default to a null cache configuration)
- Run 2 policies with the `set-labels` action

For example:

```suggestion
        p = self.load_policy(
            {
                'name': 'label-gke-cluster',
                'resource': 'gcp.gke-cluster',
                'filters': [{'name': name}],
                'actions': [{'type': 'set-labels',
                            'labels': {'test_label': 'test_value'}}]},
            cache=True,
            config=Config.empty(
                cache='memory',
                cache_period=10,
                output_dir=self.get_temp_dir(),
            ),
            session_factory=factory
        )
        resources = p.run()
        self.assertEqual(len(resources), 1)
        p = self.load_policy(
            {
                'name': 'relabel-gke-cluster',
                'resource': 'gcp.gke-cluster',
                'filters': [{'name': name}],
                'actions': [{'type': 'set-labels',
                            'labels': {'test_label': 'new_value'}}]},
            cache=True,
            config=Config.empty(
                cache='memory',
                cache_period=10,
                output_dir=self.get_temp_dir(),
            ),
            session_factory=factory
        )
        resources = p.run()
        self.assertEqual(len(resources), 1)
```

**Review by ajkerrigan — COMMENTED** [review]
Thanks for the work on this! It looks good to me and works in my local tests :tada: . It sets the stage for more resources to get label support too, which is great!

A few comments inline, most of which could probably be ignored or deferred to refactoring PRs. And I'd want :eyes: from other maintainers with better GCP instincts before merging. The only higher level concern I have is around filtering out all deprecation warnings.

File: pyproject.toml
Comment (ajkerrigan) [reply pyproject.toml:80]: Fair enough :+1:

**Comment by kapilt** [comment]
We can probably extract some common behavior but the apis aren't uniform across services wrt to fingerprint or self link usage afaicr

File: tools/c7n_gcp/c7n_gcp/actions/core.py
Comment (ajkerrigan) [reply tools/c7n_gcp/c7n_gcp/actions/core.py:88]: Nevermind this, sounds like it's not an option.

**Comment by ksonney** [comment]
> 



> We can probably extract some common behavior but the apis aren't uniform across services wrt to fingerprint or self link usage afaicr

Putting this in the correct reply. 

Basically to echo what @kapilt said below - the APIs aren't uniform in GCP wrt fingerprints and self-links. Since I'm going to be adding this for more services, this will be getting refactored as I move forward, since I expect almost all of them will require their own handling. How about I open an issue around that, and we track the refactor and addition of labels to other services there?

**Comment by ajkerrigan** [comment]
> > 
> 
> 
> 
> > We can probably extract some common behavior but the apis aren't uniform across services wrt to fingerprint or self link usage afaicr
> 
> Putting this in the correct reply. 
> 
> Basically to echo what @kapilt said below - the APIs aren't uniform in GCP wrt fingerprints and self-links. Since I'm going to be adding this for more services, this will be getting refactored as I move forward, since I expect almost all of them will require their own handling. How about I open an issue around that, and we track the refactor and addition of labels to other services there? 

That's unfortunate but not surprising given how we've had to special case so much with AWS :-/. I'm going to just watch this one from here on, apologies for the sidetracks.

File: tools/c7n_gcp/c7n_gcp/actions/core.py
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/actions/core.py:76]: I think we need to move this logic to a subclass for gke, perhaps having this base class delegate to a handle_exception hook, which defaults to raising.

File: tools/c7n_gcp/c7n_gcp/actions/core.py
Comment (kapilt) [reply tools/c7n_gcp/c7n_gcp/actions/core.py:88]: I think the instinct here is right re separating the type specific logic out of the base class with,  but a subclass with the base delegating exception handling to a method should suffice, given non uniformity around self link, fingerprint behavior.

File: tools/c7n_gcp/c7n_gcp/actions/core.py
Comment (ksonney) [reply tools/c7n_gcp/c7n_gcp/actions/core.py:76]: I don't disagree, but that's a bit beyond my knowledge of python (right now). I DID clean it up so that the individual models now have the refetchFingerprints method that allows us to define it per-model now and no have the logic for "what kind of resource is this" - if the model supports the "refetchFingerprint" function, it gets run, otherwise it raises the parent exception

Comment (ksonney) [reply tools/c7n_gcp/c7n_gcp/actions/core.py:76]: Looking at the future: to add label support to other resources, the changes needed would be: enable labels and add refetchFingerprint function to the resource, then test.

**Comment by kapilt** [comment]
looks pretty good to me, It looks like we'll need to get the test coverage up a little bit, but else I think its good. two minors, most of the method names we have our doing name_understcore instead of nameUnderscore re replaceFingerprint, and the date serialization change to handle sets, doesn't look like its used.

File: tools/c7n_gcp/c7n_gcp/resources/compute.py
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/compute.py:61]: hmm.. not sure why this is being flagged as not covered, the tests do cover afaics.

File: tools/c7n_gcp/c7n_gcp/resources/compute.py
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/compute.py:293]: if its just being reraised, don't need to wrap in try/except block

Comment (kapilt) [reply tools/c7n_gcp/c7n_gcp/resources/compute.py]: removing those will help increase coverage, and the caller is already wrapped in a try except block.

File: tools/c7n_gcp/c7n_gcp/actions/core.py
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/actions/core.py:89]: re why noqa, ruff was complaining about e not being defined without this, flake8 was happy, so a bug in ruff afaics.

**Review by kapilt — APPROVED** [review]
lgtm, thanks!

**Comment by ksonney** [comment]
Do we need to tag the `refetch_fingerprint` functions with `#noqa` since they get tested but codecov is ignoring that?

**Comment by kapilt** [comment]
it does look like coverage reporting is correct, in that at least for the compute resources, the refetch methods are never called.

**Comment by kapilt** [comment]
okay, the changes should cover on test for refresh methods for instance and image, so this will hopefully pass on coverage now.

**Pull request merged**
